bug: --single-user mode breaks every hour for non-first accounts — refresh-path rebind check

## Bug Description

In `--single-user` mode with multiple credential files, any account other than the one whose session got bound first silently breaks with "ACTION REQUIRED: Authentication Needed" **exactly one hour after its token was last refreshed**. The refresh itself works — the new access token is successfully fetched from Google and written to the credential file on disk — but the subsequent `OAuth21SessionStore.store_session()` call fails the immutable session-binding check, the caller treats that as "no valid credentials," and the user sees a re-auth prompt they don't actually need.

This is related to but distinct from #433, #506, and #642, which all describe initial call-time mismatches (first-file-alphabetical, case sensitivity, wrong email). This is the **refresh path**: everything is correct at call time, but the hourly background refresh triggers the rebind check on accounts beyond the first.

## Symptoms

In a long-lived MCP session that uses multiple `user_google_email` values in `--single-user` mode:

- **First account called** in the session: works indefinitely (refresh stays within the session binding).
- **Every other account**: works on the first call (initial access token is still valid). ~1 hour later, the next call fails with:
  > ACTION REQUIRED: Google Authentication Needed for Google Gmail for 'accountB@example.com'

The user is forced to re-approve OAuth for accountB. After re-approval, the clock resets and the same failure recurs an hour later. Filing this as Spencer Reed @spencer-c-reed — my agent traced the bug over about 30 minutes of log analysis after repeated re-auth prompts that "kept happening" over multiple days.

## Steps to Reproduce

1. Configure `--single-user` mode with credentials for two Google accounts, both with valid refresh tokens in `~/.workspace-mcp/credentials/` (default `WORKSPACE_MCP_CREDENTIALS_DIR`).
2. Start the workspace-mcp server (e.g., `workspace-mcp --tools gmail --transport streamable-http --single-user`).
3. In a single MCP client session, call `search_gmail_messages(user_google_email="accountA@example.com", ...)`. Works.
4. Call `search_gmail_messages(user_google_email="accountB@example.com", ...)`. Works (access token still valid).
5. Wait until accountB's access token expires (1 hour after it was last refreshed).
6. Call `search_gmail_messages(user_google_email="accountB@example.com", ...)` again. **Fails with "ACTION REQUIRED: Authentication Needed"**, despite accountB's refresh_token being valid and the refresh itself having succeeded.

## Log Evidence

From `~/.workspace-mcp/logs/combined.log` on v1.19.0 (accounts redacted):

```
[TOOL] [search_gmail_messages] accountB@example.com -> gmail/v1
[OAUTH] [get_credentials] Session user accountA@example.com doesn't match requested accountB@example.com; skipping session store
[OAUTH] [get_credentials] Single-user mode: bypassing session mapping, finding any credentials
[OAUTH] [get_credentials] Single-user mode: found credentials for requested user accountB@example.com
[OAUTH] [get_credentials] Credentials not valid. Attempting refresh. User: 'accountB@example.com'
[OAUTH] [get_credentials] Credentials refreshed successfully. User: 'accountB@example.com'
[CREDS] Stored credentials for accountB@example.com to /home/.../credentials/accountB@example.com.json
[ERROR] SECURITY: Attempt to rebind session <session_id> from accountA@example.com to accountB@example.com
[OAUTH] [get_credentials] Error refreshing credentials: Session <session_id> is already bound to a different user.
[OAUTH] [search_gmail_messages] No valid credentials. Email: 'accountB@example.com'.
[OAUTH] [start_auth_flow] Initiating auth for Google Gmail for 'accountB@example.com' with scopes for enabled tools.
[TOOL] [search_gmail_messages] Auth failed for accountB@example.com | method=mcp_session_binding
```

Key lines:

- "Credentials refreshed successfully" — the refresh actually worked; Google returned a new access token.
- "Stored credentials for accountB" — the new token was saved to disk.
- "SECURITY: Attempt to rebind session from accountA to accountB" — the `store.store_session()` call in `get_credentials()` then tries to update the OAuth21SessionStore with accountB's new binding, which fails because the session was already immutably bound to accountA on the first call.
- The caller treats the `ValueError` from `store_session()` as a generic refresh failure and returns `None`, which the service decorator translates into "re-auth required."

## Root Cause

`auth/oauth21_session_store.py` `store_session()` enforces an immutable session-to-user binding (lines 400–414 in v1.19.0):

```python
if mcp_session_id:
    if mcp_session_id not in self._session_auth_binding:
        self._session_auth_binding[mcp_session_id] = user_email
    elif self._session_auth_binding[mcp_session_id] != user_email:
        logger.error(
            f"SECURITY: Attempt to rebind session {mcp_session_id} from "
            f"{self._session_auth_binding[mcp_session_id]} to {user_email}"
        )
        raise ValueError(
            f"Session {mcp_session_id} is already bound to a different user"
        )
```

This is enforced on **every** `store_session()` call that passes a non-None `mcp_session_id`, including the call made inside `get_credentials()` after a successful refresh (v1.19.0 `auth/google_auth.py` around line 994):

```python
# Also update OAuth21SessionStore
store = get_oauth21_session_store()
store.store_session(
    user_email=user_google_email,
    ...
    mcp_session_id=session_id,
    ...
)
```

The binding semantics are inconsistent with single-user mode's other behavior:

- `get_credentials()` **reads** in single-user mode explicitly "bypass session mapping" (v1.19.0 `auth/google_auth.py` line 880: "Single-user mode: bypassing session mapping, finding any credentials").
- But **writes** still enforce the immutable binding, which in single-user mode has no security meaning — there is only one operator, and the credential-store lookup is what actually authenticates the caller.

The rebind check is a correct defense in multi-user HTTP-server mode, where session binding is how you prevent one user's session from being reused to act as another. In single-user mode, it's a source of false-positive failures with no corresponding security benefit.

## Suggested Fix

Drop `mcp_session_id` at the entry of `store_session()` when running in single-user mode. Centralizing the fix in `oauth21_session_store.py` covers all four call sites in `google_auth.py` (lines 151, 755, 832, 994 in v1.19.0) plus the two call sites inside `oauth21_session_store.py` itself (lines 873, 992).

```python
# auth/oauth21_session_store.py, at the top of OAuth21SessionStore.store_session()
import os  # add to file imports

def store_session(self, user_email, access_token, refresh_token=None, ..., mcp_session_id=None, issuer=None):
    """..."""
    # In --single-user mode, never create an immutable session-to-user binding.
    # Single-user mode's read path (get_credentials) explicitly bypasses session
    # mapping and looks up by user_email directly. Enforcing an immutable binding
    # on writes causes every hourly refresh of any account other than the first
    # to fail with "SECURITY: rebind" -> surfaced to the user as a spurious
    # re-auth prompt.
    if os.getenv("MCP_SINGLE_USER_MODE") == "1" and mcp_session_id is not None:
        mcp_session_id = None

    with self._lock:
        # ... existing logic
```

This is one line of guard plus one import. It preserves the rebind security check for multi-user mode (where it's load-bearing) and silently disables it for single-user mode (where it never was load-bearing and only creates false positives).

I've applied this locally and verified with an artificial 1-hour-expired token that:

- The refresh path completes cleanly for any account.
- No "SECURITY: rebind" errors appear in the log.
- All accounts continue to work across the 1-hour expiry boundary indefinitely.
- Multi-user behavior is unchanged (the `os.getenv("MCP_SINGLE_USER_MODE")` guard is false in multi-user mode).

## Environment

- **workspace-mcp version:** 1.19.0 (installed via `pip`)
- **Transport:** `streamable-http` on localhost:8000 (pm2-managed, single-user flag set)
- **Python:** 3.12
- **OS:** Ubuntu 24.04 (VPS)
- **Google accounts configured:** 3 credential files in `~/.workspace-mcp/credentials/`
- **Client:** Claude Code CLI

## Related Issues

- #433 (closed): multi-account session binding causing first-account use for all calls. Different symptom — that was a read-side issue; this is a write-side issue on the refresh path.
- #506 (closed): first-file-alphabetical mismatch. Different cause — this occurs when the correct file is loaded but the refresh still fails.
- #642 (open): caller-provided `user_google_email` override. Different condition — this occurs even when the caller passes the correct email matching the credential file.
- #561 (closed): "Token not auto-refreshed in --single-user stdio mode". Partially related — in this bug the refresh itself does succeed, but the store call after the refresh is what fails.
- #703 (closed): OAuth state validation in single-user. Different path.

Happy to open a PR with the patch and regression test if that's useful.

--- a/auth/oauth21_session_store.py
+++ b/auth/oauth21_session_store.py
@@ -587,6 +587,13 @@ class OAuth21SessionStore:
             mcp_session_id: FastMCP session ID to map to this user
             issuer: Token issuer (e.g., "https://accounts.google.com")
         """
+        # Single-user mode reads credentials directly by email and bypasses the
+        # session mapping. Keeping the immutable MCP session binding enabled here
+        # turns a successful token refresh for a second account into a false
+        # "rebind" error.
+        if os.getenv("MCP_SINGLE_USER_MODE") == "1":
+            mcp_session_id = None
+
         with self._lock:
             normalized_expiry = _normalize_expiry_to_naive_utc(expiry)
 

--- a/gmail/gmail_tools.py
+++ b/gmail/gmail_tools.py
@@ -2269,9 +2269,7 @@ def _format_thread_content(
     for i, message in enumerate(messages, 1):
         payload = message.get("payload", {})
         # Extract headers
-        headers = {
-            h["name"]: h["value"] for h in payload.get("headers", [])
-        }
+        headers = {h["name"]: h["value"] for h in payload.get("headers", [])}
 
         sender = headers.get("From", "(unknown sender)")
         date = headers.get("Date", "(unknown date)")

--- a/tests/auth/test_oauth21_session_store.py
+++ b/tests/auth/test_oauth21_session_store.py
@@ -1,3 +1,5 @@
+import pytest
+
 from auth.oauth21_session_store import OAuth21SessionStore
 
 
@@ -86,3 +88,44 @@ def test_deserialize_oauth_state_entry_normalizes_invalid_and_naive_timestamps(
     assert deserialized["created_at"] is not None
     assert deserialized["created_at"].tzinfo is not None
     assert deserialized["expires_at"] is None
+
+
+def test_store_session_rejects_mcp_session_rebind_by_default(tmp_path):
+    state_file = tmp_path / "oauth_states.json"
+    store = OAuth21SessionStore(oauth_state_file=str(state_file))
+
+    store.store_session(
+        user_email="account-a@example.com",
+        access_token="token-a",
+        mcp_session_id="session-123",
+    )
+
+    with pytest.raises(ValueError, match="already bound to a different user"):
+        store.store_session(
+            user_email="account-b@example.com",
+            access_token="token-b",
+            mcp_session_id="session-123",
+        )
+
+
+def test_store_session_skips_mcp_binding_in_single_user_mode(
+    tmp_path, monkeypatch
+):
+    monkeypatch.setenv("MCP_SINGLE_USER_MODE", "1")
+
+    state_file = tmp_path / "oauth_states.json"
+    store = OAuth21SessionStore(oauth_state_file=str(state_file))
+
+    store.store_session(
+        user_email="account-a@example.com",
+        access_token="token-a",
+        mcp_session_id="session-123",
+    )
+    store.store_session(
+        user_email="account-b@example.com",
+        access_token="token-b",
+        mcp_session_id="session-123",
+    )
+
+    assert store.get_user_by_mcp_session("session-123") is None
+    assert store.get_credentials("account-b@example.com").token == "token-b"

--- a/tests/auth/test_oauth21_session_store.py
+++ b/tests/auth/test_oauth21_session_store.py
@@ -108,9 +108,7 @@ def test_store_session_rejects_mcp_session_rebind_by_default(tmp_path):
         )
 
 
-def test_store_session_skips_mcp_binding_in_single_user_mode(
-    tmp_path, monkeypatch
-):
+def test_store_session_skips_mcp_binding_in_single_user_mode(tmp_path, monkeypatch):
     monkeypatch.setenv("MCP_SINGLE_USER_MODE", "1")
 
     state_file = tmp_path / "oauth_states.json"

